### DISCUSSION TITLE
chapter_data_structure/character_encoding/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#66] [Announcements]
# chapter_data_structure/character_encoding/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_data_structure/character_encoding/

**Comment by limuzi9806** [comment]
大佬！Fig. UTF-8 编码示例

图中的UTF-8 h和e是不是写反了？

**Reply by krahets** [reply]
是的，谢谢！

**Comment by chenjunhit** [comment]
在3.4.5中，为什么说 UTF-16是等长编码呢，它不是使用2个或者4个字符来表示字符吗~

**Reply by krahets** [reply]
是的，这个很大一部分是历史原因，以前（包括现在）大部分常用字符用 2 bytes 来表示就够了，因此称 UTF-16 是等长编码。
Java 等语言通过代理对的方式来存储 4 字符的 UTF-16 编码，此时会丧失等长编码的优势，已经属于变长编码了。

**Comment by szgz** [comment]
这样说来，python3是处理字符串最高效的？

**Reply by krahets** [reply]
这个得具体看处理哪种字符串。不过总体来说，越“新”的语言考虑的越全面，因为历史包袱少

**Comment by youngledo** [comment]
说到字符编码，我最痛恨的就是关系型数据库的字符串类型，还需要不停的设置长度，这么多年了，依旧如此！

**Comment by dlcnwoini** [comment]
懂了

**Comment by Floating-Y** [comment]
感觉还是没太明白为什么`UTF-16`可以随机访问，编码这块都有点晕的感觉😵

**Reply by krahets** [reply]
对于等长编码，假设长度为 `x` ，那么第 `i` 个元素的起始索引应该是 `(i-1) * x` 。如果是变长编码，就无法轻易地计算出每个元素的位置了，需要遍历，使用 $O(n)$ 时间。

**Reply by Floating-Y** [reply]
我明白了，谢谢k神

**Comment by CodingGorit** [comment]
提个建议，图片可以加个点击预览放大的功能，有的图内容要素过多，放大预览效果可能更好

**Reply by krahets** [reply]
好的，已加～

**Comment by newcakes** [comment]
我听不懂该怎末办

**Reply by krahets** [reply]
坚持看坚持听，慢慢会好的

**Reply by Hu-Said** [reply]
你当然听不懂，因为你只能看懂...

**Comment by cocowise** [comment]
上面关于Python 3 的内容貌似有误。借助GPT做了试验：

=====

以下是一些超出BMP范围的汉字示例：

𨊻（U+2A2BB，属于补充平面）
𨋾（U+2A2BE，属于补充平面）

创建一个 Unicode 字符串
```python
mystr = "你好，>𨊻𨋾<！"
```

导入 chardet 模块查看字符串编码
```python
import chardet
```

使用 chardet 模块检测字符串的编码
```python
result = chardet.detect(str)
```

打印字符串的编码
```python
print(result['encoding'])
```
输出：utf-8

在 Python 3 中，所有字符串都是 Unicode 字符串，默认编码是 UTF-8。

-----

以上是Bart的输出结果，可再查证确认。

**Reply by cocowise** [reply]
更正 应为 result = chardet.detect(mystr)

**Reply by krahets** [reply]
chartdet 用于检测一段“编码”的编码类型，因此需要输入 bytearray 而非字符串，例如：

```python
# data 是一串 UTF-8 编码
data = b"\xe4\xbd\xa0\xe5\xa5\xbd"
result = chardet.detect(data)
print(result)
```

**Reply by krahets** [reply]
Python 中的 `str` 字符串直接用 Unicode 表示（请注意区分：读写文件默认为 UTF-8 编码）。CPython 为 str 对象使用了一种灵活的存储策略，字符长度取决于字符串中最大的字符。观察以下测试：

```python
import sys

eng1 = "a"
eng2 = "ab"
eng3 = "abc"

chn1 = "哈"
chn2 = "哈啰"
chn2_eng1 = "哈啰a"

bmp1 = "𨊻"
bmp2 = "𨊻𨋾"
bmp2_eng1 = "𨊻𨋾a"

print("\n英文：") # 英文长度为 1
print(eng1 + ": ", sys.getsizeof(eng1))
print(eng2 + ": ", sys.getsizeof(eng2))
print(eng3 + ": ", sys.getsizeof(eng3))

print("\n中文:")
print(chn1 + ": ", sys.getsizeof(chn1))
print(chn2 + ": ", sys.getsizeof(chn2))
print(chn2_eng1 + ": ", sys.getsizeof(chn2_eng1))

print("\n补充平面:")
print(bmp1 + ": ", sys.getsizeof(bmp1))
print(bmp2 + ": ", sys.getsizeof(bmp2))
print(bmp2_eng1 + ": ", sys.getsizeof(bmp2_eng1))
```

可以得到以下结论：

1. 纯英文字符串的字符长度为 1 ；
2. 中文字符串下，中文和英文长度都为 2 ；
3. 补充平面字符下，英文字符长度为 4 ；

```python
英文：
a:  50
ab:  51
abc:  52

中文:
哈:  76
哈啰:  78
哈啰a:  80

超平面:
𨊻:  80
𨊻𨋾:  84
𨊻𨋾a:  88
```

**Reply by cocowise** [reply]
UTF-8 是变长编码，字符长度为 1、2、4是UTF-8的设定

**Reply by krahets** [reply]
@cocowise 不一样的。例如，Python 存储中英文混合字符串时，中、英文字符长度都是 2 。但如果用 UTF-8 编码它，英文字符长度为 1 ，中文字符长度为 3 （假设都是常见字符）。

**Reply by UselessWater** [reply]
确实，我刚开始也以为python中的编码的说法和utf-8一样，为啥还要特意说是灵活编码，然后看到了评里面的解释，我才恍然大悟，差点误解了。

**Comment by serenNan** [comment]
那c++是用什么编码呢，顺便想问一下现在用c++的人很少了吗，因为现在不知道主学什么语言😢

**Reply by SayhiUnreal** [reply]
https://cloud.tencent.com/developer/article/2350500?areaId=106001

**Comment by aeilang** [comment]
一直被这些编码困惑，零星地搜索材料也是一知半解，没想到作者这么系统又简洁地解释！这种风格让我联想起了Golang的风格

**Comment by limit4ever** [comment]
一直有个疑问，为什么vscode使用utf-8编码写c/cpp的代码中出现中文就会乱码，但是用gbk编码就不会

**Reply by krahets** [reply]
是不是 Windows ？应该是因为 Windows 中文默认采用 GBK 编码。

**Reply by xiong-ZH-zq** [reply]
Windows控制台输出中文默认采用GBK编码

**Reply by HuaWang135608** [reply]
VSCode 的解决办法可以看我的博客：[VSCode运行C/C++代码控制台输出中文的乱码问题](https://www.cnblogs.com/huawang135608/p/14851532.html)

**Reply by wxi7641** [reply]
eclipse也是这样，代码中出现中文需要用gbk编码，否则就会乱码

**Comment by 5710xIa** [comment]
最常见的莫过于`锟斤拷`了，这时候觉得秦王统一文字是多么大的功劳。

**Comment by Keekuun** [comment]
建议搞个pwa，可以安装到桌面

**Comment by HC-miss** [comment]
😁很棒！！！

**Comment by will9701** [comment]
这么多年终于搞懂了。。。

**Comment by liuyd96** [comment]
很少给别人写评论，但这一章确实惊艳到我了，非常非常详细，真的是意外惊喜

**Comment by Lnncoco** [comment]
> UTF-16 编码：使用 2 或 4 字节来表示一个字符。
> ...程序运行中的字符串都采用 UTF-16 或 UTF-32 这类等长编码。

按前面描述，UTF-16应该是属于变长编码？

**Reply by krahets** [reply]
Hi 原始的 UTF-16 是 2 字节表示一个字符。而为了表示更多字符，它采用“代理对”形成 4 字节。

**Comment by 480284856** [comment]
图3.8那里的“算”和“法”位啥变成三个字节表示了呀，没搞明白呢😂

**Reply by krahets** [reply]
因为这是 UTF-8 编码的规定

**Reply by elon-lo** [reply]
在ASCII码的范围，用一个字节表示，超出ASCII码的范围就用字节表示，这就形成了我们上面看到的UTF-8的表示方法，这样的好处是当UNICODE文件中只有ASCII码时，存储的文件都为一个字节，所以就是普通的ASCII文件无异，读取的时候也是如此，所以能与以前的ASCII文件兼容。
大于ASCII码的，就会由上面的第一字节的前几位表示该unicode字符的长度，比如110xxxxx前三位的二进制表示告诉我们这是个2BYTE的UNICODE字符；1110xxxx是个三位的UNICODE字符，依此类推；xxx的位置由字符编码数的二进制表示的位填入。越靠右的x具有越少的特殊意义。只用最短的那个足够表达一个字符编码数的多字节串。注意在多字节串中，第一个字节的开头"1"的数目就是整个串中字节的数目。
例：如果为两个字节，表示方法为 110xxxxx 10xxxxxx

**Reply by elon-lo** [reply]
三个字节，表示方法为 1110xxxx 10xxxxxx 10xxxxxx
四个字节，表示方法为 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx 10xxxxxx
以此类推
以上内容来自维基百科查询，https://zh.wikipedia.org/wiki/UTF-8

**Reply by 2510466299** [reply]
k神，上面你提到在'python'中，其存储中英文混合字符串时，中、英文字符长度都是 2 ，请问为什么书中’hello 算法‘的’算法‘是三个字节表示？

**Reply by 2510466299** [reply]
书中提到中文的常用字符是三个字节表示，但中英混合的字符串中中文是两个字符表示，由此我对于这个例子’hello 算法‘产生了一些疑惑。

**Reply by krahets** [reply]
@2510466299 Hi，应注意区分：

> 书中提到中文的常用字符是三个字节表示

这个介绍的是 UTF-8 编码的方式。

> 在'python'中，其存储中英文混合字符串时，中、英文字符长度都是 2 ，

这个是 Python 代码中的字符编码方式，其不是 UTF-8 编码。

**Reply by 2510466299** [reply]
好的，多谢k神解惑！

**Comment by deyu9999** [comment]
使用某种编程语言写代码时，IDE中看得见的代码字符按照UTF-8编码后被存储在硬盘中；
当程序需要run的时候，看得见的代码符号经过编译器，按照“编程语言的编码方式”Unicode的编译映射，字符串放置到了内存堆栈中。
以上是对“字符串在编程语言中的存储方式”的理解。

**Comment by zmvw235** [comment]
But why not use a specific character to represent the coding length of the following characters? When there are few alterations, I think it may be a better solution.

**Reply by zmvw235** [reply]
Sorry it's hard to type Chinese on my raspberry so I use English

**Reply by krahets** [reply]
Everything in memory will be stored as a binary number, so "use a specific character to represent the coding length" will pose a new recursive problem: how to encode that specific character in binary?

**Comment by NgxuAnGMH** [comment]
- UTF-8 的通用性最佳，许多工具和库优先支持 UTF-8

说实话编程语言的字符编码这个...以前从来没有了解过，真是学艺不精了。

**Comment by ysssgdhr** [comment]
讲的很清楚👍

**Comment by huih99** [comment]
确实写的很详细，大概搞懂了各种编码了

**Comment by Wwh-SoEximious** [comment]
学计算机两年了，终于搞懂了。我很少称赞别人，这章真是惊艳到我了，简洁明了，逻辑清晰👍

**Comment by Wwh-SoEximious** [comment]
文中提到的“代理对”是什么呀？🤔

**Comment by floatDreamWithSong** [comment]
编码这一篇写的很清晰，大有收获😋

**Comment by zhishiheng** [comment]
为什么 UTF-8表示中文用3个字节，UTF-16只用2个字节。既然2个字节够用，为什么要用三个字节？

**Reply by zhishiheng** [reply]
因为10占位了

**Reply by wuditjdx** [reply]
8个占位了吧，4+2+2

**Reply by wyang0824** [reply]
unicode是全球通用，英文一个字节就够了，用utf16也多占用了内存，utf8更灵活

**Comment by Daylotve77** [comment]
这一章讲得太清楚了，真很值得一看

**Comment by luoyaocray** [comment]
the best articles of coding

**Comment by Aka-Edward** [comment]
他奶奶的孙子的！好乱！脑子疼要长脑子了！

**Comment by 490273789** [comment]
之前也查过相关的知识，看的云里雾里，看着这篇文章突然的就懂了。

**Comment by WangFengtu1996** [comment]
示例解析（以汉字“算”的Unicode码点U+7B97为例）：
Unicode二进制：0111 1011 1001 0111
UTF-8编码步骤：
首字节：取高4位0111，填充到1110xxxx → 11100111（十六进制0xE7）
第二字节：取中间6位101110，填充到10xxxxxx → 10101110（十六进制0xAE）
第三字节：取低6位010111，填充到10xxxxxx → 10010111（十六进制0x97）
最终编码为E7 AE 97

**Comment by HatsuneMikuSama** [comment]
utf-16中如何判断四字节是由两个2字节组成的还是仅仅就是一个4字节的字符？

**Comment by Pengyu-Jin** [comment]
讲得很棒，感谢分享

**Comment by pyk0627** [comment]
厉害

**Comment by Lingxuan-Ye** [comment]
> 在庞大的 Unicode 字符集中，常用的字符占用 2 字节，有些生僻的字符占用 3 字节甚至 4 字节。

unicode 是字符与码点的映射，码点是个纯数字概念，不存在某个字符占用多少字节的说法。比如，某个字符的码点为 1，那么这个 1 仅仅只是数字 1，自然数 0 的后继；用几个字节表示、用中文还是英文表示，用什么进制表示，都与其本质无关。

相应的，utf-8 等才是编码方案，才可以说在这个编码下某个字符占用多少字节。

**Reply by krahets** [reply]
谢谢指正！已修复

**Comment by believebelief** [comment]
难

**Comment by DeanYao2014** [comment]
Unicode = 字符集（定义字符与码点）,UTF‑8 = 编码（定义码点如何转成字节）;  以前根们就没有字符集和编码的概念...

**Comment by DeanYao2014** [comment]
这个unicode是字符集 和码点一一对应, 通常16进制只时便于展示;  
utf-8是编码规则,文档/ html保存的都是utf8 的字节, 如果要显示那么必须要utf8 解码 得到 码点 , 码点一一对应映射unicode字符集.
UTF‑8 文本是一串字节流。解码时从左到右扫描：看首字节的前导 1 的数量 → 知道这个字符占几个字节 → 取出这些字节 → 拼成码点 → 查 Unicode 字符集 → 得到字符。
